File: DelSole.CrossAnalyzers.VisualBasic/DelSole.CrossAnalyzers.VisualBasic/AsyncSuffixAnalyzer.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class AsyncSuffixAnalyzer
    Inherits DiagnosticAnalyzer

    Public Const DiagnosticId = "CRA010"
    Friend Shared ReadOnly Title As LocalizableString = "Names of asynchronous methods should end with Async"
    Friend Shared ReadOnly MessageFormat As LocalizableString = "Name of asynchronous method '{0}' does not end with Async"
    Friend Const Category = "Naming"

    Friend Shared Rule As New DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, True,
                                                   helpLinkUri:="https://github.com/AlessandroDelSole/DotNetAnalyzers/wiki/DNA-200---Names-of-asynchronous-methods-should-end-with-Async")

    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Rule)
        End Get
    End Property

    Public Overrides Sub Initialize(context As AnalysisContext)
        ' TODO: Consider registering other actions that act on syntax instead of or in addition to symbols
        context.RegisterSymbolAction(AddressOf AnalyzeSymbol, SymbolKind.Method)
    End Sub

    Private Sub AnalyzeSymbol(context As SymbolAnalysisContext)
        ' TODO: Replace the following code with your own analysis, generating Diagnostic objects for any issues you find

        Dim methodSymbol = CType(context.Symbol, IMethodSymbol)

        If methodSymbol.IsAsync = False Then
            Return
        End If

        If methodSymbol.HandledEvents.Any Then
            Return
        End If

        If Not methodSymbol.Name.ToLowerInvariant.EndsWith("async") Then
            Dim diag = Diagnostic.Create(Rule, methodSymbol.Locations(0), methodSymbol.Name)

            context.ReportDiagnostic(diag)
        End If
    End Sub
End Class

File: DelSole.CrossAnalyzers.CSharp/DelSole.CrossAnalyzers.CSharp/AsyncSuffixAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace DelSole.CrossAnalyzers.CSharp
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class AsyncSuffixAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "CRA002";
        static internal readonly LocalizableString Title = "Names of asynchronous methods whould end with Async";
        static internal readonly LocalizableString MessageFormat = "Name of asynchronous method '{0}' does not end with Async";
        internal const string Category = "Naming";
        static internal DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, true, helpLinkUri: "https://github.com/AlessandroDelSole/DelSole.CrossAnalyzers/wiki");

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.Method);
        }


        private void AnalyzeSymbol(SymbolAnalysisContext context)
        {
            // TODO: Replace the following code with your own analysis, generating Diagnostic objects for any issues you find

            var methodSymbol = (IMethodSymbol)context.Symbol;

            if (methodSymbol.IsAsync == false)
            {
                return;
            }

            if (!methodSymbol.Name.ToLowerInvariant().EndsWith("async"))
            {
                var diag = Diagnostic.Create(Rule, methodSymbol.Locations[0], methodSymbol.Name);

                context.ReportDiagnostic(diag);
            }
        }
    }
}